README.md
---
# IBM Fully Homomorphic Encryption Toolkit for Linux

The IBM Fully Homomorphic Encryption (FHE) Toolkit for Linux is packaged as [Docker][1] containers that make it easier to get started and experimenting with the [Fully Homomorphic Encryption][2] technology.

This repository contains all the scripts required to install and run the pre-packaged toolkits as a docker container. The toolkits comprise the [IBM Homomorphic Encryption Library - HElib][3], an [Integrated Development Environment (IDE)][4] and the corresponding workspace containing the ready-to-run example code in a variety of supported toolkit editions. The editions suported are based on [Centos][5], [Fedora][6], or [Ubuntu][7]. Each toolkit edition provides access to the built in IDE via a web browser on your host. For the more adventurous users, instructions to build and package the toolkit locally are also provided.

If you are instead looking for the IBM Fully Homomorphic Encryption Tookit for macOS/iOS that provide a native toolkit for Apple developers, it can be found <a href="https://github.com/IBM/fhe-toolkit-macos" target="_blank">here</a>.

## Supported Configurations

At this time, the toolkits support many <a href="https://www.docker.com/resources/what-container" target="_blank">Docker</a> capable hosts such as most modern Linux distributions, macOS, <a href="https://docs.microsoft.com/en-us/windows/wsl/install-win10" target="_blank">Windows 10 Subsystem for Linux</a> and <a href="https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.4.0/com.ibm.zos.v2r4.izso100/izso100_whatisintro.htm" target="_blank">z/OS Container Extensions</a>. Other host operating systems with recent Docker software may work as well, but are untested. On x86-64 based systems we support all three operating system editions of the toolkit. For the s390x architecture, the toolkit presently supports the Ubuntu and Fedora Editions. We hope to offer other s390x operating system options for the toolkits soon. 


## Running the Toolkit

If you want to dive right in and get started installing a pre-built container available from [Docker Hub][8] and using the pre-configured IDE Workspace, please see the [Getting Started Document](GettingStarted.md). 

## Building the Toolkit

If you want to build and deploy the toolkit from scratch and perhaps customize the container to your needs, please see the [Getting Started Advanced Document](GettingStarted.Advanced.md).


## Feedback Survey
 
"IBM invites you to participate in our Advanced Security and Encryption Survey. This survey is designed to gather insights around the security challenges you or your organization face and better understand our users and how to serve you better. We will only use your feedback to improve the FHE Toolkit experience and inform future IBM security-focused products and services. IBM will not share your response data with any third parties. We look forward to your valuable feedback to improve the IBM Fully Homomorphic Encryption Toolkit for macOS, iOS, and Linux."

https://www.surveygizmo.com/s3/5731822/Advanced-Security-And-Encryption-Survey-2020


## About this distribution

The remainder of this README file explains how the source code and project are organized for those who might want to contribute to, or modify, the toolkit.  

This toolkit installs the IDE and extensions required to compile, build and run the included HElib based examples and any application you choose to write to exploit the homomorphic encryption technology. The end result is a containerized environment complete with the source code editor, making it easy for you to develop applications using HElib.

To preserve user privacy, automatic telemetry and data collection by the IDE have been disabled in this distribution.

Docker is required to build the containerized toolkit. If you are new to Docker, we recommend starting with [Docker Desktop][9]. 

## Source Code Overview

This repository contains a set of Dockerfiles, scripts, and IDE configuration files to run the IBM FHE Toolkit for Linux. The toolkit runs in a Linux docker container and uses the IDE as a workspace and development environment accessible through a browser in your host system.

The toolkit is built from one of the following parent images depending on the platform specified to the setup script:

- fedora:32
- centos:8
- ubuntu:20.04

The same setup script also downloads [HElib][3] and and a required dependency Number Theory Library ([NTL][10]). These are stored in `<directory where you cloned the toolkit>/DEPENDENCIES` in the host machine. 

The HElib and NTL distributions are built and installed globally in `/usr/local` as a shared library in the toolkit docker container making them available to include in your program. HElib and NTL also depend on the [GNU Multiple Precision Arithmetic Library (GMP)][11], which is pre-installed when building the docker container.

The demo application source code is copied from `HElib/examples` into a workspace directory `/opt/IBM/FHE-Workspace`. The HElib source distribution can be found in `/opt/IBM/FHE-distro/HElib`. Both may be accessed through the IDE workspace running in the toolkit container, accessed through the web browser on your desktop. Always use the workspace when trying to work with any of these components.   


   [1]: https://www.docker.com/                                  "Docker Container"
   [2]: https://en.wikipedia.org/wiki/Homomorphic_encryption     "Homomorphic Encryption"
   [3]: https://github.com/homenc/HElib/                         "HElib"
   [4]: https://code.visualstudio.com/                           "Visual Studio Code"
   [5]: https://www.centos.org/                                  "CentOS"
   [6]: https://getfedora.org/                                   "Fedora"
   [7]: https://ubuntu.com/                                      "Ubuntu"
   [8]: https://hub.docker.com/u/ibmcom                          "Docker Hub IBM"
   [9]: https://www.docker.com/get-started/                      "Docker get started"
   [10]: https://www.shoup.net/ntl/                               "NTL"
   [11]: https://gmplib.org/                                     "GMP library"



CHANGES.md
IBM Fully Homomorphic Encryption Toolkit for Linux 1.0.2.0
Public Release
======
(tagged as v1.0.2.0)

## IBM FHE Toolkit Core Team
Co-authored-by: Flavio Bergamaschi <flavio@uk.ibm.com> <br/>
Co-authored-by: Greg Boland <boland@us.ibm.com> <br/>
Co-authored-by: Emma Dawson <emmadawson@uk.ibm.com> <br/>
Co-authored-by: Eli M Dow <emdow@us.ibm.com> <br/>
Co-authored-by: Rosie Lickorish <rosie.lickorish@uk.ibm.com> <br/>

## IBM FHE HElib Team
Co-authored-by: Flavio Bergamaschi <flavio@uk.ibm.com> <br/>
Co-authored-by: Jack Crawford <jack.crawford@ibm.com> <br/>
Co-authored-by: Hamish Hunt <hamishun@uk.ibm.com> <br/>
Co-authored-by: Enrico Steffinlongo <enrico.steffinlongo@ibm.com> <br/>

## IBM ML-HElib Team
Co-authored-by: Allon Adir <adir@il.ibm.com> <br/>
Co-authored-by: Ehud Aharoni <aehud@il.ibm.com> <br/>
Co-authored-by: Gilad Ezov <gilad.ezov@ibm.com> <br/>
Co-authored-by: Dov Murik <dov.murik1@il.ibm.com> <br/>
Co-authored-by: Omri Soceanu <omri.soceanu@il.ibm.com> <br/>

## Collaborators
Co-authored-by: Brian Fernandez <brian.fernandez@ibm.com> <br/>
Co-authored-by: Güvenc Gülce <Guevenc.Guelce@ibm.com> <br/>
Co-authored-by: Brandon Johnson <branjohn@us.ibm.com> <br/>
Co-authored-by: Norhan Khaled <norhan.ali.araba@ibm.com> <br/>
Co-authored-by: Anthony Moran <amoran@us.ibm.com> <br/>
Co-authored-by: Pradeep Parameshwaran <pradeep@de.ibm.com> <br/>


RunToolkit.sh
#!/bin/bash

# MIT License
# 
# Copyright (c) 2020 International Business Machines
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

HElib_version='v1.0.2'

# We default to persistent storage because this is a code development toolkit and developers likely will make code changes and expect that to live on...
SANDBOX_MODE=0

# We default to running a container pulled from Docker Hub
CONTAINER_MODE=""

# Set default port for docker to use
PORT=8443

# The default location on this host where a project directory of FHE toolkit source code will be persisted
DEFAULT_FHE_PERSISTENT_WORKSPACE_PATH="$PWD"/FHE-Toolkit-Workspace
zCX=$(docker system info | grep platform)
if [ $zCX  ] && [ $zCX == "platform=zOS" ]; then
  DEFAULT_FHE_PERSISTENT_WORKSPACE_PATH=AZD_SHARED_VOLUME
fi

# Formatting helpers
bold=$(tput bold)
normal=$(tput sgr0)

#
BASEDIR="$PWD"/$(dirname $0)
SCRIPTNAME=$(basename $0)

print_usage(){
  cat <<EOF

${bold}Usage: $scriptname <-p|-l> [options] CONTAINER_OS${normal}

${bold}CONTAINER_OS${normal}   Selects the name of the operating system platform for
               the FHE Toolkit container you want to use.
               Available OS are:
               x86_64/amd64: {ubuntu, fedora, centos}
               s390x:        {ubuntu}

-p             Run the IBMCOM pre-built toolkit from Docker Hub

-l             Run the locally built toolkit

${bold}Options:${normal}
-v <directory> Specify a nonstandard directory for the persisted workspace.
               By default the IDE workspace will be persisted in
               $DEFAULT_FHE_PERSISTENT_WORKSPACE_PATH

-s             Run the toolkit in stateless mode without a persisted workspace.

-h             Displays this help information.


EOF

}



#echo "$@"

# The number of parameters passed to this script
NPARAM=$#

while getopts ":hpslv:" opt; do
  case ${opt} in
    h ) # Usage
      print_usage
      exit 0
      ;;
    v ) # User wants a persistent volume. No sandbox. 
      FHE_WORKSPACE_PATH=$OPTARG
      SANDBOX_MODE="0"
      ;;
    s ) # User wants an ephemeral sandbox. No peristent volume.
      SANDBOX_MODE="1"
      ;;
    l ) # Run locally built contaliner
      [ -n "$CONTAINER_MODE" ] \
      && { echo "Fatal: Options -l and -p are mutually exclusive" ; print_usage; exit -1;} \
      || CONTAINER_MODE=local
      ;;
    p ) # Run ibmcom pre-built container
      [ -n "$CONTAINER_MODE" ] \
      && { echo "Fatal: Options -l and -p are mutually exclusive" ; print_usage; exit -2; } \
      || CONTAINER_MODE=ibmcom
      ;;
    \? ) # Usage
      print_usage
      exit 0
      ;;
    : ) # Invalid option. Print usage
      echo "Invalid option: -$OPTARG requires an argument" 1>&2
      print_usage
      exit -3
      ;;
  esac
done

shift $((OPTIND -1))
#echo "$@"

OPTINDNUMBER=$((OPTIND-1))

platform=$1

if [[ $((NPARAM - OPTINDNUMBER)) -ne 1 ]]; then
  echo "Fatal: Options and parameters mismatch or missing parameters. Please check order and number of parameters."
  print_usage
  exit -4
fi



# Determine which container distributuon to launch
if ! [ "$platform" ]
then
  echo "Missing OS platform name - Please specify a supported platform"
  print_usage
  exit -5
elif [[ "$platform" = "fedora" ]]
then
    platform="fedora"
    tag="32"
elif [[ "$platform" = "centos" ]]
then
    platform="centos"
    tag="8"
elif [[ "$platform" = "ubuntu" ]]
then
    platform="ubuntu"
    tag="20.04"
else
    echo "Invalid option: $platform - Please specify a supported platform"
    print_usage
    exit -6
fi

# We need to use absolute paths for the -v volume option, and users like to use relative paths, so make it work for them...
function abs_path {
  local dir=`dirname "$1"`
  local file=`basename "$1"`
  pushd "$dir" &>/dev/null || return $? # On error, return error code
  echo "`pwd -P`/$file" # output full, link-resolved path with filename
  popd &> /dev/null
}


# Here we handle user options with respect to ephemeral or persistent code storage within the toolkit.
# Define an empty string or a valid docker volume option to contain the users intent
DOCKER_PERSISTENT_VOLUME_ARGS=""
# If we are running persistent mode, respect the users invocation...
if [ "$SANDBOX_MODE" == "0" ]; then 
  if [ ! "${FHE_WORKSPACE_PATH}" ]; then
    FHE_WORKSPACE_PATH="${DEFAULT_FHE_PERSISTENT_WORKSPACE_PATH}"
    echo "INFO:    Using system default persistent storage path..."
  else
    echo "INFO:    Using user-specified persistent storage path..."
  fi
    
  # Convert to absolute path to protect against relative paths unless on zCX
  if [ ! $zCX ]; then
  FHE_WORKSPACE_PATH=$(abs_path "$FHE_WORKSPACE_PATH")
  fi
  DOCKER_PERSISTENT_VOLUME_ARGS="-v ${FHE_WORKSPACE_PATH}:/opt/IBM/FHE-Workspace "
  echo "INFO:    Persistent data storage: \"${FHE_WORKSPACE_PATH}"\"

  # Now we have some kind of Persistent path to use, we should check that it is sane (writeable)
  # Note we check against the unescaped path version
  if [ ! -w "$FHE_WORKSPACE_PATH" ] && [ ! $zCX ]; then
    echo "FATAL:   The requested persistent path \"$FHE_WORKSPACE_PATH\" does not exist and cannot be used for persistent storage of the fhe toolkit runtime."
    echo "         Please check that this directory exists and contains a valid project workspace. One can use the provided PersistData.sh script to set up a location."
    exit -7
  fi

  # When using the default path, make sure the cmake configuraion is cleaned each time
  if [ "$FHE_WORKSPACE_PATH" == "$DEFAULT_FHE_PERSISTENT_WORKSPACE_PATH" ]; then
    echo "INFO:    CMake: Deleting cached built settings and reconfigure"
    CMAKE_CACHE=$DEFAULT_FHE_PERSISTENT_WORKSPACE_PATH/build/CMakeCache.txt
    if [[ -f "$CMAKE_CACHE" ]]; then
      rm $CMAKE_CACHE
    fi
    CMAKE_FILES=$DEFAULT_FHE_PERSISTENT_WORKSPACE_PATH/build/CMakeFiles
    if [[ -d "$CMAKE_FILES" ]]; then
      rm -r $CMAKE_FILES
    fi
  fi

  pattern=" |'|\"|!"
  if [[ $FHE_WORKSPACE_PATH =~ $pattern ]]; then
    echo "FATAL:   The requested persistent path \"$FHE_WORKSPACE_PATH\" must not contain spaces or some special characters."
    echo "         Please check that this directory path does not contain spaces or any special characters or non printable characters"
    exit -8
  fi

# If we are in ephemeral mode we can leave the docker volume arg string entirely blank to be ignored at docker run time
else
  :
  echo "INFO:    User-specified sandbox mode with no persistent data storage"
  # The container will use the included file system that is ephemeral for the example in the well known path /opt/IBM/HE-Workspace/
fi

# Check for LOCAL vs Docker Hub build choice
if [ ${CONTAINER_MODE} == "local" ]; then
  FHEkit_image_name=local/fhe-toolkit-${platform}
  FHEkit_container_name=local-fhe-toolkit-${platform}
elif [ ${CONTAINER_MODE} == "ibmcom" ]; then
  FHEkit_image_name=ibmcom/fhe-toolkit-${platform}
  FHEkit_container_name=fhe-toolkit-${platform}
else
  echo "FATAL: Container mode $CONTAINER_MODE is invalid"
  print_usage
  exit -9
fi

#exit 0

CID=$(docker ps -q -f status=running -f name=fhe)
if [ ! "${CID}" ]; then
  # Container is not running...

  FHE_WORKSPACE_PATH=$( echo "${FHE_WORKSPACE_PATH}" | sed 's/ /\\ /g' )
  
  cmd="docker run -d --name $FHEkit_container_name  $DOCKER_PERSISTENT_VOLUME_ARGS -p $PORT:8443 $FHEkit_image_name"

  echo "INFO:    Launching FHE tookit: "
  echo ""
  echo ""
  echo "         $cmd"
  echo ""
  echo ""
  $cmd
  
  DOCKER_RET_CODE=$?
  if [ $DOCKER_RET_CODE -ne "0" ]; then
      echo ""
      echo ""
      echo "****************************************************************************"
      echo "FATAL:    Docker container launch failed. Aborting."
      echo "          Make sure you run the FetchDockerImage.sh script for pre-packaged"
      echo "          images or the BuildDockerIMage.sh script for locally built images."
      echo "****************************************************************************"
      echo ""
      echo ""
      exit -10
  fi 
  # If we made it here print the info about the new running container...
  echo ""
  docker ps -a --filter name=fhe-toolkit-$platform
  echo "" 
  echo "FHE Development is open for business: https://127.0.0.1:$PORT/"
else
  echo ""
  echo ""
  echo "*****************************************************************************************************"
  echo "WARNING: An existing instance already running. Only a single FHE toolkit instance may run presently. "
  echo "         The instance must be stopped before a new toolkit instance can be launched:" 
  echo "*****************************************************************************************************"
  echo "" 
  echo "" 
  docker ps -a -f name=fhe-
  echo "" 
fi
unset CID


